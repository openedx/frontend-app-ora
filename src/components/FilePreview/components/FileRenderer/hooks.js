import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
