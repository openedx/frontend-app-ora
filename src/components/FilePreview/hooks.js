import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
