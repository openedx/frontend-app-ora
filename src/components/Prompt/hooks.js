import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
