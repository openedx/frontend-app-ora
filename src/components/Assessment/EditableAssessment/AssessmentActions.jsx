import React from 'react';
import PropTypes from 'prop-types';

import { Button, StatefulButton } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useCloseModal } from 'hooks/modal';
import { MutationStatus } from 'constants';

import { useViewStep } from 'hooks/routing';
import messages, { viewStepMessages } from '../messages';

/**
 * <Rubric />
 */
const AssessmentActions = ({
  onSubmit,
  submitStatus,
}) => {
  const closeModal = useCloseModal();
  const { formatMessage } = useIntl();
  const step = useViewStep();
  const viewStep = viewStepMessages[step] ? `${formatMessage(viewStepMessages[step])} ` : '';

  return (
    <div className="assessment-footer">
      <Button className="w-100" onClick={closeModal} variant="outline-primary">
        {formatMessage(messages.finishLater)}
      </Button>
      <StatefulButton
        className="w-100"
        onClick={onSubmit}
        state={submitStatus}
        labels={{
          [MutationStatus.idle]: formatMessage(messages.submitGrade, { viewStep }),
          [MutationStatus.loading]: formatMessage(messages.submittingGrade),
          [MutationStatus.success]: formatMessage(messages.gradeSubmitted),
        }}
      />
    </div>
  );
};

AssessmentActions.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  submitStatus: PropTypes.string.isRequired,
};

export default AssessmentActions;
