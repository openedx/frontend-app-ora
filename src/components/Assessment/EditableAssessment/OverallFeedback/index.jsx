import React from 'react';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useViewStep } from 'hooks/routing';
import {
  useOverallFeedbackPrompt,
  useOverallFeedbackFormFields,
} from 'hooks/assessment';

import InfoPopover from 'components/InfoPopover';
import messages from 'components/Assessment/messages';

import { stepNames } from 'constants/index';

/**
 * <OverallFeedback />
 */
const OverallFeedback = () => {
  const { formatMessage } = useIntl();
  const prompt = useOverallFeedbackPrompt();
  const { value, onChange } = useOverallFeedbackFormFields();

  const step = useViewStep();
  if (step === stepNames.studentTraining) {
    return null;
  }
  return (
    <Form.Group>
      <Form.Label className="criteria-label">
        <span className="criteria-title">
          {formatMessage(messages.overallComments)}
        </span>
        <InfoPopover>
          <div data-testid="prompt-test-id">{prompt}</div>
        </InfoPopover>
      </Form.Label>
      <Form.Control
        as="textarea"
        className="rubric-feedback feedback-input"
        floatingLabel={formatMessage(messages.addComments)}
        value={value}
        onChange={onChange}
      />
    </Form.Group>
  );
};

export default OverallFeedback;
