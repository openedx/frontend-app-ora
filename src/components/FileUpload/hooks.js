import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useDownloadFiles } from 'hooks/app';

export const stateKeys = StrictDict({
  shouldShowError: 'shouldShowError',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});

export const useUploadConfirmModalHooks = ({
  file,
  closeHandler,
  uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, '');
  const [shouldShowError, setShouldShowError] = useKeyedState(
    stateKeys.shouldShowError,
    false,
  );

  const confirmUploadClickHandler = () => {
    // console.log({ confirmUploadClick: { description } });
    if (description !== '') {
      uploadHandler(file, description);
    } else {
      setShouldShowError(true);
    }
  };

  const exitHandler = () => {
    setShouldShowError(false);
    setDescription('');
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    shouldShowError,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({ onFileUploaded }) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(
    stateKeys.isModalOpen,
    false,
  );

  const confirmUpload = useCallback(async (file, description) => {
    // console.log({ confirmUpload: { file, description } });
    setIsModalOpen(false);
    if (onFileUploaded) {
      // console.log({ uploadArgs });
      await onFileUploaded({ ...uploadArgs, description });
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(
    ({ fileData, handleError, requestConfig }) => {
      setIsModalOpen(true);
      setUploadArgs({ fileData, handleError, requestConfig });
    },
    [setIsModalOpen, setUploadArgs],
  );

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export const useFileDownloadHooks = ({ files, zipFileName }) => {
  const downloadFileMation = useDownloadFiles();
  const downloadFiles = () => downloadFileMation.mutate({
    files,
    zipFileName,
  });
  return {
    downloadFiles,
    status: downloadFileMation.status,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
  useFileDownloadHooks,
};
