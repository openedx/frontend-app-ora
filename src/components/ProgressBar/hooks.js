import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
