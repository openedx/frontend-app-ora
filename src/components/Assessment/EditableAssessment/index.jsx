import React from 'react';

import { Card } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useCriteriaConfig } from 'hooks/assessment';

import CriterionContainer from 'components/CriterionContainer';
import RadioCriterion from 'components/CriterionContainer/RadioCriterion';
import CriterionFeedback from 'components/CriterionContainer/CriterionFeedback';

import OverallFeedback from './OverallFeedback';
import AssessmentActions from './AssessmentActions';

import messages from '../messages';

/**
 * <Rubric />
 */
const EditableAssessment = () => {
  const { formatMessage } = useIntl();

  const criteria = useCriteriaConfig();

  const criteriaContainers = criteria.map(
    (criterion, criterionIndex) => (
      <CriterionContainer
        key={criterion.name}
        criterion={criterion}
        input={<RadioCriterion {...{ criterion, criterionIndex }} />}
        feedback={<CriterionFeedback {...{ criterion, criterionIndex }} />}
      />
    ),
  );

  return (
    <Card className="assessment-card">
      <Card.Section className="assessment-body">
        <h3>{formatMessage(messages.rubric)}</h3>
        <hr className="m-2.5" />
        {criteriaContainers}
        <hr />
        <OverallFeedback />
      </Card.Section>
      <AssessmentActions />
    </Card>
  );
};

EditableAssessment.propTypes = {
};

export default EditableAssessment;
