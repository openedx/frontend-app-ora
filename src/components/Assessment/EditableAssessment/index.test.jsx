import { shallow } from '@edx/react-unit-test-utils';

import { useCriteriaConfig } from 'hooks/assessment';
import EditableAssessment from '.';

jest.mock('hooks/assessment', () => ({
  useCriteriaConfig: jest.fn(),
}));

jest.mock('components/CriterionContainer', () => 'CriterionContainer');
jest.mock('components/CriterionContainer/RadioCriterion', () => 'RadioCriterion');
jest.mock('components/CriterionContainer/CriterionFeedback', () => 'CriterionFeedback');
jest.mock('./OverallFeedback', () => 'OverallFeedback');
jest.mock('./AssessmentActions', () => 'AssessmentActions');

describe('<EditableAssessment />', () => {
  it('render empty criteria', () => {
    useCriteriaConfig.mockReturnValue([]);
    const wrapper = shallow(<EditableAssessment />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('CriterionContainer')).toHaveLength(0);
  });

  it('render with criteria', () => {
    const mockCriteria = [
      { name: 'criterion1' },
      { name: 'criterion2' },
      { name: 'criterion3' },
    ];
    useCriteriaConfig.mockReturnValue(mockCriteria);
    const wrapper = shallow(<EditableAssessment />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('CriterionContainer')).toHaveLength(mockCriteria.length);
  });
});
