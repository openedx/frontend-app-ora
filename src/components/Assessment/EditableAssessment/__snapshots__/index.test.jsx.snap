// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableAssessment /> render empty criteria 1`] = `
<Card
  className="assessment-card"
>
  <Card.Section
    className="assessment-body"
  >
    <h3>
      Rubric
    </h3>
    <hr
      className="m-2.5"
    />
    <hr />
    <OverallFeedback />
  </Card.Section>
  <AssessmentActions />
</Card>
`;

exports[`<EditableAssessment /> render with criteria 1`] = `
<Card
  className="assessment-card"
>
  <Card.Section
    className="assessment-body"
  >
    <h3>
      Rubric
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      criterion={
        Object {
          "name": "criterion1",
        }
      }
      feedback={
        <CriterionFeedback
          criterion={
            Object {
              "name": "criterion1",
            }
          }
          criterionIndex={0}
        />
      }
      input={
        <RadioCriterion
          criterion={
            Object {
              "name": "criterion1",
            }
          }
          criterionIndex={0}
        />
      }
      key="criterion1"
    />
    <CriterionContainer
      criterion={
        Object {
          "name": "criterion2",
        }
      }
      feedback={
        <CriterionFeedback
          criterion={
            Object {
              "name": "criterion2",
            }
          }
          criterionIndex={1}
        />
      }
      input={
        <RadioCriterion
          criterion={
            Object {
              "name": "criterion2",
            }
          }
          criterionIndex={1}
        />
      }
      key="criterion2"
    />
    <CriterionContainer
      criterion={
        Object {
          "name": "criterion3",
        }
      }
      feedback={
        <CriterionFeedback
          criterion={
            Object {
              "name": "criterion3",
            }
          }
          criterionIndex={2}
        />
      }
      input={
        <RadioCriterion
          criterion={
            Object {
              "name": "criterion3",
            }
          }
          criterionIndex={2}
        />
      }
      key="criterion3"
    />
    <hr />
    <OverallFeedback />
  </Card.Section>
  <AssessmentActions />
</Card>
`;
